ros1_ign_point_cloud/src/point_cloud.cc

```diff
@@ -0,0 +1,347 @@
+  }
+  if (_format != "FLOAT32")
+  {
+    ROS_WARN_NAMED("ros1_ign_point_cloud",
+        "Expected depth image to have [FLOAT32] format, but it has [%s]", _format.c_str());
+  }
+
+  // Fill message
+  // Logic borrowed from
+  // https://github.com/ros-simulation/gazebo_ros_pkgs/blob/kinetic-devel/gazebo_plugins/src/gazebo_ros_depth_camera.cpp
+  auto sec_nsec = ignition::math::durationToSecNsec(this->current_time_);
+
+  sensor_msgs::PointCloud2 msg;
+  msg.header.frame_id = this->frame_id_;
+  msg.header.stamp.sec = sec_nsec.first;
+  msg.header.stamp.nsec = sec_nsec.second;
+  msg.width = _width;
+  msg.height = _height;
+  msg.row_step = msg.point_step * _width;
+  msg.is_dense = true;
+
+  sensor_msgs::PointCloud2Modifier modifier(msg);
+  modifier.setPointCloud2FieldsByString(2, "xyz", "rgb");
+  modifier.resize(_width*_height);
+
+  sensor_msgs::PointCloud2Iterator<float> iter_x(msg, "x");
+  sensor_msgs::PointCloud2Iterator<float> iter_y(msg, "y");
+  sensor_msgs::PointCloud2Iterator<float> iter_z(msg, "z");
+  sensor_msgs::PointCloud2Iterator<uint8_t> iter_rgb(msg, "rgb");
+
+  if (this->rgb_camera_)
+  {
+    this->rgb_camera_->Capture(this->rgb_image_);
+    fillImage(this->rgb_image_msg_, sensor_msgs::image_encodings::RGB8, _height,
+        _width, 3 * _width, this->rgb_image_.Data<unsigned char>());
+  }
+
+  double hfov = this->depth_camera_->HFOV().Radian();
+  double fl = _width / (2.0 * tan(hfov / 2.0));
+  int index{0};
+  uint8_t * image_src = (uint8_t *)(&(this->rgb_image_msg_.data[0]));
+
+  // Convert depth to point cloud
+  for (uint32_t j = 0; j < _height; ++j)
+  {
+    double p_angle;
+    if (_height>1)
+      p_angle = atan2( (double)j - 0.5*(double)(_height-1), fl);
+    else
+      p_angle = 0.0;
+
+    for (uint32_t i=0; i<_width; i++, ++iter_x, ++iter_y, ++iter_z, ++iter_rgb)
+    {
+      double y_angle;
+      if (_width>1)
+        y_angle = atan2( (double)i - 0.5*(double)(_width-1), fl);
+      else
+        y_angle = 0.0;
+
+      double depth = _scan[index++];
+
+      // in optical frame
+      // hardcoded rotation rpy(-M_PI/2, 0, -M_PI/2) is built-in
+      // to urdf, where the *_optical_frame should have above relative
+      // rotation from the physical camera *_frame
+      *iter_x = depth * tan(y_angle);
+      *iter_y = depth * tan(p_angle);
+      if (depth > this->depth_camera_->FarClipPlane())
+      {
+        *iter_z = ignition::math::INF_D;
+        msg.is_dense = false;
+      }
+      if (depth < this->depth_camera_->NearClipPlane())
+      {
+        *iter_z = -ignition::math::INF_D;
+        msg.is_dense = false;
+      }
+      else
+      {
+        *iter_z = depth;
+      }
+
+      // Put image color data for each point
+      // \TODO(anyone) RGB image seems offset from depth image by 3~4 pixels in both directions
+      if (this->rgb_image_msg_.data.size() == _height * _width * 3)
+      {
+        // color
+        // \TODO(anyone) why do we need to do BHR instead of RGB?
```

Comment:
When calling `modifier.setPointCloud2FieldsByString(2, "xyz", "rgb");` its creating the rgb field as 3 unsigned chars stacked in a float. Accessing the bytes of the float directly is probably why the colors are backward because they are using little endian byte order. The documentation [here](http://docs.ros.org/melodic/api/sensor_msgs/html/namespacesensor__msgs.html) suggests creating `PointCloud2Iterator`s for each color channel.

Reply:
Niiiice! Fixed in 7eede34757687e87f494d7450f7837be4142b09b